src/frontend/flight/sync/utils.rs

```diff
@@ -414,6 +415,288 @@ pub(super) fn get_prune_map(
+        let expr = match upper_sync_col.role() {
+            ColumnRole::Changed => {
+                project_changed_column(upper_sync_col, lower_schema, SyncPosition::Upper)?
+            }
+            ColumnRole::Value => project_value_column(
+                upper_sync_col,
+                upper_schema,
+                lower_schema,
+                &mut col_desc,
+                &mut projection,
+                SyncPosition::Upper,
+            )?,
+            _ => continue, // We already picked up PKs from the lower sync
+        };
+
+        col_desc.insert(upper_sync_col.column_descriptor());
+        projection.push(expr);
+    }
+
+    let col_desc = col_desc.into_iter().collect::<Vec<_>>();
+
+    // Finally use a qualified alias for the projection, which will act as a new lower sync in the
+    // subsequent sync merge
+    let projection = col_desc
+        .iter()
+        .zip(projection)
+        .map(|(cd, expr)| {
+            expr.alias_qualified(Some(LOWER_SYNC), SyncColumn::canonical_field_name(cd))
+        })
+        .collect();
+
+    Ok((col_desc, projection))
+}
+
+enum SyncPosition {
+    Lower,
+    Upper,
+}
+
+fn project_old_pk_column(
+    lower_old_pk_col: &SyncColumn,
+    upper_schema: &SyncSchema,
+) -> SyncResult<Expr> {
+    let lower_old_pk_field = lower_old_pk_col.field().name();
+
+    let upper_old_pk_col = upper_schema
+        .column(lower_old_pk_col.name(), ColumnRole::OldPk)
+        .expect("Old Pk columns must be the same across syncs");
+
+    // For un-matched lower sync rows use lower sync old PKs, otherwise
+    // for un-matched upper sync rows use upper sync old PKs, otherwise
+    // for matched lower-upper sync rows use lower sync old PKs
+    Ok(when(
+        col(UPPER_SYNC).is_null(),
+        col(Column::new(Some(LOWER_SYNC), lower_old_pk_field)),
+    )
+    .otherwise(
+        when(
+            col(LOWER_SYNC).is_null(),
+            col(Column::new(
+                Some(UPPER_SYNC),
+                upper_old_pk_col.field().name(),
+            )),
+        )
+        .otherwise(col(Column::new(Some(LOWER_SYNC), lower_old_pk_field)))?,
+    )?)
+}
+
+fn project_new_pk_column(
+    lower_new_pk_col: &SyncColumn,
+    upper_schema: &SyncSchema,
+) -> SyncResult<Expr> {
+    let lower_new_pk_field = lower_new_pk_col.field().name();
+
+    let upper_new_pk_col = upper_schema
+        .column(lower_new_pk_col.name(), ColumnRole::NewPk)
+        .expect("Old Pk columns must be the same across syncs");
+
+    // For un-matched lower sync rows use lower sync new PKs, otherwise
+    // for un-matched and matched upper sync rows use upper sync new PKs
+    Ok(when(
+        col(UPPER_SYNC).is_null(),
+        col(Column::new(Some(LOWER_SYNC), lower_new_pk_field)),
+    )
+    .otherwise(col(Column::new(
+        Some(UPPER_SYNC),
+        upper_new_pk_col.field().name(),
+    )))?)
+}
+
+fn project_changed_column(
```

Comment:
I feel like the merged upper/lower routine now have too much stuff going on, and are actually detrimental to understanding, will split them up.